component/local/file/detector.go

```diff
@@ -0,0 +1,204 @@
+	// DetectorFSNotify uses filesystem events to wait for changes to the file.
+	DetectorFSNotify
+	// DetectorPoll will re-read the file on an interval to detect changes.
+	DetectorPoll
+
+	// DetectorDefault holds the default UpdateType.
+	DetectorDefault = DetectorFSNotify
+)
+
+var (
+	_ encoding.TextMarshaler   = Detector(0)
+	_ encoding.TextUnmarshaler = (*Detector)(nil)
+)
+
+// String returns the string representation of the UpdateType.
+func (ut Detector) String() string {
+	switch ut {
+	case DetectorFSNotify:
+		return "fsnotify"
+	case DetectorPoll:
+		return "poll"
+	default:
+		return fmt.Sprintf("Detector(%d)", ut)
+	}
+}
+
+// MarshalText implements encoding.TextMarshaler.
+func (ut Detector) MarshalText() (text []byte, err error) {
+	return []byte(ut.String()), nil
+}
+
+// UnmarshalText implements encoding.TextUnmarshaler.
+func (ut *Detector) UnmarshalText(text []byte) error {
+	switch string(text) {
+	case "":
+		*ut = DetectorDefault
+	case "fsnotify":
+		*ut = DetectorFSNotify
+	case "poll":
+		*ut = DetectorPoll
+	default:
+		return fmt.Errorf("unrecognized detector %q, expected fsnotify or poll", string(text))
+	}
+	return nil
+}
+
+type fsNotify struct {
+	opts    fsNotifyOptions
+	watcher *fsnotify.Watcher
+	cancel  context.CancelFunc
+}
+
+type fsNotifyOptions struct {
+	Logger      log.Logger
+	Filename    string
+	UpdateCh    chan<- struct{} // Where to send detected updates to
+	RewatchWait time.Duration   // How often to try to re-watch the file
+}
+
+// newFSNotify creates a new fsnotify detector which uses filesystem events to
+// detect that a file has changed.
+func newFSNotify(opts fsNotifyOptions) (*fsNotify, error) {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	if err := w.Add(opts.Filename); err != nil {
+		// It's possible that the file already got deleted by the time our fsnotify
+		// was created. We'll log the error and wait for our polling fallback for
+		// the file to be recreated.
+		level.Warn(opts.Logger).Log("msg", "failed to watch file", "err", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	wd := &fsNotify{
+		opts:    opts,
+		watcher: w,
+		cancel:  cancel,
+	}
+
+	go wd.wait(ctx)
+	return wd, nil
+}
+
+func (fsn *fsNotify) wait(ctx context.Context) {
+	rewatchTick := time.NewTicker(fsn.opts.RewatchWait)
+	defer rewatchTick.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-rewatchTick.C:
+			// The fsnotify watcher may have removed our file from the watch list
+			// (i.e., if the file got deleted).
+			//
+			// Continually re-adding it to the watcher acts as a fallback polling
+			// mechanism, where we'll eventually rewatch the file once it gets added
+			// again.
+			err := fsn.watcher.Add(fsn.opts.Filename)
+			if err != nil {
+				level.Warn(fsn.opts.Logger).Log("msg", "failed re-watch file", "err", err)
+			} else {
+				fsn.forwardNotification()
+			}
+		case err := <-fsn.watcher.Errors:
```

Comment:
Under what circumstances will we receive errors here? Why do we want to forwardNotification() and read the file here, so that it fails and the output gets cleared?

Reply:
Looks like the errors you could get depends on the platform, but the intent was to force the file to reload so any file-level issues would be reported as a health problem on the component. 

Do you think this should take a different approach or just needs more comments to explain what's going on?

Reply:
No, I just wanted to clarify the logic here. Thanks!

Reply:
Gotcha, I'll add comments then :) If it wasn't obvious from your first read, it won't be obvious to others either and they shouldn't have to find the PR to look for an explanation 👍